Asssertions with Reference Data

U8NWXD opened the pull request:
For many processes, it is difficult to write simple assertions. For
example, it is not easy to write an assertion that a growth curve is as
expected. Instead of explicitly describing the curves we expect (e.g. as
an equation), we can save reference data that describes expected
behavior and then compare output to that reference. This PR
introduces the tools to:

* Save reference timeseries as CSV files
* Load reference timeseries from CSV files
* Compare timeseries to check that they are sufficiently correlated. We
  check for correlation instead of equality to allow for handling of
  stochastic processes.

As an example, this PR also implements these assertions for `convenience_kinetics` and adds associated reference data.

U8NWXD commented:
@eagmon I was thinking about the reference data implementation we discussed this afternoon, so I created a draft implementation to see what it would look like.

@eagmon @prismofeverything let me know what you think!

eagmon commented on vivarium/compartment/composition.py:
```diff
@@ -2,8 +2,14 @@
 import matplotlib.pyplot as plt
-
-from vivarium.utils.dict_utils import deep_merge, deep_merge_check
+import pandas as pd
```
Is pandas used here only for reading/writing csvs?  We don't currently have it in our pyenv and requirements, so maybe you can use the standard python csv library?  Unless you plan to use it for anything else.

U8NWXD replied:
Yes, pandas is only used for the CSV part right now. It [looks like](https://stackoverflow.com/a/10373268) this is easy to achieve with `csv`, so I'll switch over to that.

Just FYI: I noticed that `pandas` is already required by `cobra`, as you can see when installing the packages from the requirements file:
```
Collecting pandas>=0.17.0 (from cobra==0.17.1->-r requirements.txt (line 19))
```

U8NWXD replied:
I switched over to using `csv` in 1cdca73

eagmon commented on vivarium/compartment/composition.py:
```diff
@@ -371,3 +381,58 @@ def plot_simulation_output(timeseries, settings={}, out_dir='out'):
+        for col_header, row_dict in df_dict.items()
+    }
+
+def assert_timeseries_correlated(
```
What happens if the timeseries are of different lengths, or different time steps? We might also want to pass in the time array corresponding to each time series, so that they can be compared at matching time signatures.

U8NWXD replied:
It [looks like](https://stats.stackexchange.com/a/133171) correctly correlating timeseries is actually rather difficult, so I'll instead only compare the timepoints that are in both timeseries and then assert that a large enough fraction of the non-reference timeseries has been checked.

U8NWXD replied:
Implemented in ad3342d

eagmon commented on vivarium/processes/convenience_kinetics.py:
```diff
@@ -227,14 +237,25 @@ def test_convenience_kinetics(end_time=10):
     del saved_data[0]
     timeseries = convert_to_timeseries(saved_data)
     plot_simulation_output(timeseries, plot_settings, out_dir)
+    save_timeseries(timeseries, out_dir)
```
I like this -- so the ```out_dir``` gets the new timeseries csv with when the process is run.  If we like that output, we can move it over to the ```TEST_OUT_DIR``` as a permanent reference.

U8NWXD commented:
@eagmon I made changes to address the comments you left. Specifically, I removed the pandas dependency and now only compare the shared timepoints.

@prismofeverything, @eagmon I also added a function to assert that timeseries are close (differences between paired data points are at most a tolerance) since I think that better describes what we often want to check.

eagmon commented:
this looks great!
